python/ray/air/execution/barrier.py

```diff
@@ -0,0 +1,118 @@
+from typing import Any, Callable, List, Optional
+
+from ray.util.annotations import DeveloperAPI
+
+
+@DeveloperAPI
+class Barrier:
+    """Barrier to collect results and process them in bulk.
+
+    A barrier can be used to collect multiple results and process them in bulk once
+    a certain count or a timeout is reached.
+
+    For instance, if ``max_results=N``, the :meth:`on_completion` callback will be
+    invoked once :meth:`arrive` has been called ``N`` times.
+
+    ``max_results`` can be ``None``, in which case an infinite amount of results
+    will be collected. In this case, the :meth:`on_completion` callback
+    will not be invoked on successful task resolution.
+
+    Args:
+        max_results: Maximum number of results to collect before a call to
+            :meth:`wait` resolves or the :meth:`on_completion` callback is invoked.
+            If ``None``, will collect an infinite number of results.
+        complete_on_first_error: If True, will trigger completion once a single error
+            is received.
+
+    """
+
+    def __init__(
+        self, max_results: Optional[int] = None, complete_on_first_error: bool = False
+    ):
+        raise NotImplementedError
+
+    def arrive(self, *data):
+        """Notify barrier that a result successfully arrived.
+
+        This will count against the ``max_results`` limit. The received result
+        will be included in a call to :meth:`get_results`.
+
+        Args:
+            *data: Result data to be cached. Can be obtained via :meth:`get_results`.
+
+        """
+        raise NotImplementedError
+
+    def error(self, *data):
+        """Notify barrier that an error arrived.
+
+        This will count against the ``max_results`` limit. The received arguments
+        will be included in a call to :meth:`get_errors`.
+
+        If ``Barrier.complete_on_first_error`` is set, this will immediately trigger
+        the completion callback.
+
+        Args:
+            *data: Error data to be cached. Can be obtained via :meth:`get_errors`.
+        """
+        raise NotImplementedError
+
+    def on_completion(self, callback: Callable[["Barrier"], None]) -> "Barrier":
```

Comment:
I'm not sure about the different use-cases for this, but one question I had while reading the API is if it would be valuable to split this `on_completion` callback into two, with the second being a one specfically for `on_first_error`?

Reply:
I think this does make sense, I'll update the abstraction.

Reply:
Updated